메세지 도메인에 시간, 문구 정책을 추가

Bue-von-hon opened the pull request:
### 배경
#110 이슈를 해결하기 위한 첫번째 PR입니다.
메시지 정책과 문구에 대한 정책을 추가했습니다.

### 작업 내용
추가(lombok.config): 롬복이 생성해준 소스 코드는 테스트에서 제외
수정(build.gradle):
jupiter api와 테스트 엔진을 추가,
@NotEmpty 어노테이션을 위한 vallidation 의존성 추가
추가(MessageType): 메시지가 광고/정보성인지 타입에 대한 enum 클래스 생성
추가(Notification):
메시지 본문과 타이틀을 담고있는 클래스 생성,
만약 접두어가 없으면 접두어를 추가하는 메소드 생성
추가(Message):
파라미터가 추가되는 상황을 고려하여 Builder 패턴 구현,
광고성 메시지라면 접두어가 있는지 확인하고 없으면 넣어주게 구현,
광고성 메시지라면 21:00:00 ~ 08:00:59 사이에는 전송하지 못하게 구현
추가(NotificationTest): 접두어가 없으면 넣어주는 테스트 구현
추가(MessageTest):
전송 가능한 시간에 대한 테스트 추가,
정보성 메시지면 항상 전송 가능한지 테스트 추가,
광고성 메시지면 접두어를 추가해주는지 테스트 추가

### 결과
앞으로 전송 정책에 맞게 메시지를 전송할 수 있습니다.

Bue-von-hon committed:
메세지 도메인에 시간, 문구 정책을 추가

배경
#110 이슈를 해결하기 위한 첫번째 PR입니다.
메시지 정책과 문구에 대한 정책을 추가했습니다.

작업 내용
추가(lombok.config): 롬복이 생성해준 소스 코드는 테스트에서 제외
수정(build.gradle):
jupiter api와 테스트 엔진을 추가,
@NotEmpty 어노테이션을 위한 vallidation 의존성 추가
추가(MessageType): 메시지가 광고/정보성인지 타입에 대한 enum 클래스 생성
추가(Notification):
메시지 본문과 타이틀을 담고있는 클래스 생성,
만약 접두어가 없으면 접두어를 추가하는 메소드 생성
추가(Message):
광고성 메시지라면 접두어가 있는지 확인하고 없으면 넣어주게 구현
광고성 메시지라면 21:00:00 ~ 08:00:59 사이에는 전송하지 못하게 구현
추가(NotificationTest): 접두어가 없으면 넣어주는 테스트 구현
추가(MessageTest):
전송 가능한 시간에 대한 테스트 추가,
정보성 메시지면 항상 전송 가능한지 테스트 추가,
광고성 메시지면 접두어를 추가해주는지 테스트 추가

결과
앞으로 전송 정책에 맞게 메시지를 전송할 수 있습니다.

f-lab-Carrey commented on MSG-lab-common/build.gradle:
```diff
@@ -8,6 +8,12 @@ repositories {
 
+dependencies {
+    implementation 'javax.validation:validation-api:2.0.1.Final'
+    testImplementation 'org.junit.jupiter:junit-jupiter-api:5.7.2'
+    testImplementation 'org.junit.jupiter:junit-jupiter-engine:5.9.2'
```
위 test 모듈은 다른 모듈에서도 동일하게 쓰일 것 같은데요. 의존성에 대한 정의를 더 상위에서 해보면 좋겠습니다.

Bue-von-hon replied:
루트 모듈에서 주입해주도록 수정했습니다!

f-lab-Carrey commented on MSG-lab-common/src/main/java/com/example/common/domain/Message.java:
```diff
@@ -0,0 +1,61 @@
+        this.notification = notification;
+    }
+
+    public boolean isSendAble() {
```
sendable이란 단어가 합성어가 아니기에 A를 소문자로 하는게 좋을 듯 하네요

f-lab-Carrey commented on MSG-lab-common/src/main/java/com/example/common/domain/Message.java:
```diff
@@ -0,0 +1,61 @@
+        return MessageType.COMMERCIAL == type;
+    }
+
+    boolean availableTime(int hour, int minute, int second) {
+        if (require(hour < 8) || require(hour > 21)) {
+            return false;
+        }
+        if (require(hour == 8)) {
+            return require(minute > 0);
+        }
+        if (require(hour == 21)) {
+            return require(second > 0);
+        }
+        return true;
```
java8에서 제공하는 LocalTime 같은 클래스 대신 시/분/초를 이렇게 나누신 이유가 있을까요?

f-lab-Carrey commented on MSG-lab-common/src/main/java/com/example/common/domain/Message.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    boolean require(Boolean condition) {
+        return !Boolean.FALSE.equals(condition);
```
boolean에 대한 비교 시, equals를 사용하는 이유가 있을까요?

f-lab-Carrey commented on MSG-lab-common/src/test/java/com/example/common/domain/NotificationTest.java:
```diff
@@ -0,0 +1,16 @@
+    @Test
+    @DisplayName("접두어가 존재하지 않으면 넣어주는지 테스트")
+    void test1() {
+        final Notification notification = new Notification("특가 세일", "본문");
+        notification.putPrefixIfAbsent();
+        Assertions.assertTrue(notification.hasPrefix());
```
이 부분을 보면 접두어를 넣어준다는 내용이 있는데요. 
광고성 메시지와 정보성 메시지를 반별할 수 있는 기준은 무엇일까요?

f-lab-Carrey replied:
음 시스템 입장에서는 invalid한 메시지 발송 요청이 들어왔는데 임의로 (광고)라는 prefix를 붙여줬다는 생각이 드네요
사용자가 광고성 표기를 착각하여 [광고] 특가 세일이라는 문구를 발송했다면 어떨까요?
(광고)[광고] 특가세일이라고 발송되어 사용자가 의도한 용례와는 다른 메시지가 발송 될 것 같네요.

f-lab-Carrey commented on MSG-lab-common/src/main/java/com/example/common/domain/Message.java:
```diff
@@ -0,0 +1,61 @@
+
+@Getter
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
```
이 생성자들은 사용예가 없어보입니다. 필요한 생성자일까요?

f-lab-Carrey commented on MSG-lab-common/src/main/java/com/example/common/domain/Message.java:
```diff
@@ -0,0 +1,61 @@
+    private LocalDateTime time;
+
+    public Message(String to, Notification notification) {
+        id = UUID.randomUUID().toString();
+        this.to = to;
+        this.notification = notification;
```
type, time 같은 필드는 해당 생성자로는 초기화가 안되네요.
이 값들은 null이어도 괜찮을까요?

f-lab-Carrey commented on MSG-lab-common/src/main/java/com/example/common/domain/Message.java:
```diff
@@ -0,0 +1,61 @@
+        }
+
+        notification.putPrefixIfAbsent();
+        return availableTime(time.getHour(), time.getMinute(), time.getSecond());
```
return availableTime이라는 코드가 가능한 시간을 리턴해주는 것처럼 보여 boolean 타입의 리턴 타입과는 다소 괴리가 있어보입니다. 더 좋은 코드로 변경되면 좋겠습니다

Bue-von-hon committed:
작업 내용
수정(build.gradle): 의존성을 루트 모듈에서 주입, 서브 모듈에서 주입하던 의존성 삭제
수정(Notification): 메시지 타입 정보를 추가
수정(Message): 메시지 타입 정보를 Notification 클래스에 질의함, LocalTime 클래스의 메소드를 활용해 전송 가능한지 검증
수정(MessageTest): 반복적인 빌더 코드를 private 메소드로 만들어 재활용
수정(NotificationTest): Notification 클래스에 메시지 타입 정보가 추가되어 수정

Bue-von-hon reviewed (commented):
코드 리뷰를 바탕으로 수정했습니다!
p.s. 
시간/문구에 대한 전송 정책 외에 다른 정책도 추가될 수 있도록 확장 포인트를 만들고 싶습니다.
고려중인 사항으로는 필터 체인 혹은 정책을 컴포지션하는 방법이 떠오릅니다.
의견이 있다면 코멘트를 남겨주세요🙂

Bue-von-hon committed:
Fix 의존성 버전이 맞지 않아 빌드 실패하는것을 수정

작업 내용
수정(build.gradle): Declaring a dynamic version

See Also
https://docs.gradle.org/current/userguide/dependency_management.html#declaring_versions:~:text=Example%2057.%20Declaring%20a%20dependency%20with%20a%20dynamic%20version

f-lab-Carrey commented on MSG-lab-common/src/main/java/com/example/common/domain/Notification.java:
```diff
@@ -0,0 +1,36 @@
+    @NotEmpty
+    private String title;
+    private String body;
+    private MessageType type;
```
MessageType을 받는 것도 좋은 방법이지만, 사용자는 편법으로 MeesageType.INFORMATION + "(광고) 특가판매"라는 메시지를 발송할 수도 있습니다.
타입은 정보성으로 왔지만 메시지 본문은 광고성이라 괴리가 있는데요. 
이런 괴리를 어떻게 풀어보면 좋을까요?

Bue-von-hon replied:
기존의 도메인 클래스를 정보/광고 메시지 클래스로 분리했는데요.
각각의 클래스에서 메시지가 올바른지 확인해보는것도 좋을 것 같아서 추가해봤습니다! [d47883f](https://github.com/f-lab-edu/MSG-lab/pull/111/commits/d47883fd8b2811aecab55ec74ddb211a6de23521)

f-lab-Carrey commented on MSG-lab-common/src/main/java/com/example/common/domain/Message.java:
```diff
@@ -0,0 +1,50 @@
+        time = LocalDateTime.now();
+    }
+
+    private static boolean isAfterPM9(final LocalTime time) {
+        return time.isAfter(LocalTime.of(20, 59, 59));
+    }
+
+    private static boolean isBeforeAM8(final LocalTime time) {
+        return time.isBefore(LocalTime.of(8, 1, 0));
+    }
```
이 두개의 메서드는 static 메서드로 정의한 이유가 있을까요~?

Bue-von-hon replied:
인텔리제이 메소드 추출기능에서 자동으로 정적 메소드로 추출해줬습니다😅
static 키워드는 삭제했습니다!! [4f1bcf3](https://github.com/f-lab-edu/MSG-lab/pull/111/commits/4f1bcf33f196608397e2812da2084059b4e568a5)

f-lab-Carrey commented on MSG-lab-common/src/main/java/com/example/common/domain/Message.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    boolean isSendableTime(final LocalTime time) {
+        return !isBeforeAM8(time) && !isAfterPM9(time);
```
negation문이 들어가 있어서 코드를 읽을 때 조금 여러번 생각하게 되는 것 같네요 조금 더 직관적인 코드로 변경해보면 어떨까 싶습니다.

Bue-von-hon replied:
수정했습니다! [2f671c8](https://github.com/f-lab-edu/MSG-lab/pull/111/commits/2f671c8406514ca32bc36e335dae244f66d947f1)

f-lab-Carrey commented on MSG-lab-common/src/main/java/com/example/common/domain/Message.java:
```diff
@@ -0,0 +1,50 @@
+@Getter
+@Builder
+@AllArgsConstructor
+public class Message {
```
하나의 클래스에 정보성/광고성에 대한 코드가 혼잡되어 있는 느낌도 듭니다. 도메인클래스를 나눠보는 것도 방법이 될 수 있겠네요

Bue-von-hon replied:
푸시 메시지 클래스를 상속받은 정보성/광고성 메시지 클래스를 생성했습니다! [cd4bfdb](https://github.com/f-lab-edu/MSG-lab/pull/111/commits/cd4bfdb48ae3b9926b4b4261a9e23467940a18ec)
상속 관계에서는 @Builder 대신 @SuperBuilder를 상위/하위 클래스에서 사용해야해서 수정했습니다!

Bue-von-hon committed:
기존 메시지 클래스에서 정보성/광고성 푸시 메시지 클래스 추출

작업 내용
수정(checkstyle): protect 접근자 허용
삭제(message): 클래스 삭제
삭제(MessageType): 클래스 삭제
추가(PushMessage): 메시지 클래스 대신 사용할 푸시 메시지 클래스 생성
추가(CommercialPushMessage, InfromationalPushMessage): 푸시 메시지 클래스를 상속받은 도메인 클래스 생성
수정(Notification): 불필요한 코드 삭제
수정(MessageTest, PushMessageTest): 새로 생성한 클래스를 이용하도록 수정

Bue-von-hon committed:
불필요한 static 키워드 삭제

Bue-von-hon committed:
부정문 삭제

Bue-von-hon committed:
광고성 메시지이지만 정보성으로 전송되는 케이스 추가

작업 내용
수정(PushMessage): 올바른 노티피케이션인지 확인하는 추상 메소드 추가
수정(CommercialPushMessage, InformationalPushMessage): 올바른 노티피케이션인지 확인하는 추상 메소드 구현체 추가, 접두어를 통해 확인하도록 구현
수정(Notification): 광고성 접두어라는 의미가 보이도록 메소드 네임 수정
수정(MessageTest): 불필요한 캐스팅 삭제, 광고성이지만 정보성 메시지로 전송되는 테스트 케이스 추가

Bue-von-hon reviewed (commented):
코드 리뷰를 바탕으로 실험적인 구현체들을 작업해봤어요.
코멘트를 자유롭게 남겨주세요🙏

Bue-von-hon committed:
메시지 국제화, 전송 정책 판독기 생성

배경
광고성 푸시 메시지는 하단에 수신 거부 방법을 명시해야 합니다.
전송 가능 여부를 메시지 자체에서 판단하게 된다면,
메시지 클래스가 너무 많은 역할을 맡게되는 문제가 있었습니다.
수신 거부 방법에 대한 로직을 추가했습니다.
전송 가능 여부를 판단하는 판독기 클래스를 생성했습니다.
각종 문자열 메시지에 대한 국제화 기능을 추가했습니다.

작업 내용
수정(PushMessage): 전송 가능 여부에 대한 추상 메소드 삭제
수정(Notification): 접두어가 없으면 넣어주는 로직 삭제, 광고 수신 거부 방법이 본문 아래에 명시되었는지 확인하는 메소드 추가
수정(CommercialPushMessage): 전송 가능 여부 메소드 삭제, Localtime을 LocalDateTime으로 변경(국제화시 날짜가 필요하다고 판단)
수정(InformationalPushMessage): 전송 가능 여부 메소드 삭제
추가(PushMessageInfo): 광고 메시지 수신에 동의했는지 여부를 보여주는 클래스 추가
추가(PushMessageResource): 국제화를 위한 리소스 번들 클래스 추가
추가(PushMessageResource_ko_KR): 국제화를 위한 리소스 추가
추가(SendPolicyValidator): 메시지가 전송 가능한지 여부를 판단하는 판독기 추가, 광고 메시지 수신 동의 여부와 시간, 광고 접두어, 수신 거부 방법이 포함되었는지 확인하여 전송 가능 여부를 판단
수정(MessageTest): 전송 가능 여부 테스트 삭제
수정(NotificationTest): 광고 메시지 수신거부 방법이 적혀있는지 확인하는 메소드 테스트 추가, 광고 메시지의 접두어가 존재하는지 확인하는 메소드 테스트 추가
추가(SendPolicyValidator): 메시지 전송 정책에 대한 테스트 추가

결과
광고 메시지의 접두어와 광고 메시지 본문에 포함되는
수신 거부 메시지에 대한 국제화 기능을 사용할 수 있습니다.
전송 가능한지에 대한 여부를 메시지 자체가 아닌 판독기에서 판단합니다.

Bue-von-hon commented:
@f-lab-Carrey 웹서칭 중 알게된 푸시 메시지 전송 정책을 추가하고, 국제화 기능을 추가했습니다...!

참고
[메이크샵의 푸시 메시지 전송 정책](https://www.makeshop.co.kr/newmakeshop/front/faq_view.html?uid=588)
